fixes #150 Backend modified for transaction spanning across multiple wallets

olemis opened the pull request:
Refs #150

Changes:
- UXID array lookup IDs for signing now support `#<NUMBER>` relative to transaction array
- UXID array lookup IDs for signing now support absolute `<UXID>`
- Add `BlockchainTransactionAPI` interface for creating multi-wallet transactions
- Add `BlockchainSignService` interface for signing multi-wallet transactions
- New altcoin plugin entry points for instantiating aforementioned API's

Does this change need to mentioned in CHANGELOG.md?
yes

Requires testing
yes

Test cases modified to test for UXID index lookup. New test cases added for muli wallet spend behavior.

olemis committed:
[core] refs #150 - Core API classes for blockchain API

olemis committed:
[util] refs #150 - Simple reusable wallet outputs

olemis committed:
[sky] refs #150 - Rename SkycoinBlockchainStatus => SkycoinBlockchain

olemis committed:
[sky] refs #150 - Transaction building methods in SkycoinBlockchain type

- skyAPICreateTxn factored out to reuse in SkycoinBlockchain methods
- Add SimpleWalletAddress type
- Add SimpleWalletOutput type

olemis committed:
[core] refs #150 - Augment AltcoinPlugin interface with LoadTransactionAPI method

olemis committed:
[sky] refs #150 - Implement LoadTransactionAPI entry point for SkyFiber altcoin plugin

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
[sky] refs #150 - LoadTransactionAPI returns ErrInvalidNetworkType if netType is not MainNet

olemis committed:
[core] refs #150 - Add BlockchainSignService type

olemis committed:
[coin] [util] refs #150 - Add SkycoinSignService

... and reusable function implementing generic strategy for trivial sign services

olemis committed:
[sky] refs #150 - Entry point for loading sign service. Implemented by SkyFiber plugin

olemis committed:
[sky] refs #150 - Absolute (<UXID>) and relative (#<NUMBER>) input indices for signing

olemis committed:
[sky] [models] [params] refs #150 - All tickers belong in SkycoinFamily. Global constant DataRefrenshTimeout

olemis committed:
Remove file accidentally commited in 40ee5140241deebf148bca705d893417c608ad48

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
[sky] refs #150 - CalculatedHoursFamily => SkycoinFamily

... fixes error reported in https://travis-ci.org/fibercrypto/FiberCryptoWallet/jobs/606249453#L634-L638

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
refs #150 - Multi-wallet spending in CHANGELOG

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

mauricio1802 committed:
[wallets] Change Transfer method from LocalWallet to fit with the new method signature

mauricio1802 committed:
[test] Add test for SendFromAddress method from SkycoinBlockchain

mauricio1802 committed:
[test] Add test for Spend method from SkycoinBlockchain type

mauricio1802 committed:
[test] Create method generateRandomKeyData for generate random secKey, pubKey pairs from random mnemonics

mauricio1802 committed:
[test] Create makeTransactionFromMultipleWallets function

mauricio1802 committed:
[txn] Change runtime type returned by LocalWallet::signSkycoinTxn method from UninjectedTransaction to CreatedTrasaction

mauricio1802 committed:
[test] Change LocalWallet::signSkycoinTxn to return the same type that is received and Finish multisign test for createdTxn

mauricio1802 committed:
[test] Add SkycoinUninjectedTransction test in sign txns with inputs from multiple wallets

mauricio1802 committed:
[models] Add signer property to walletManager

mauricio1802 committed:
[models] Change walletManager::SendFromAddresses method to work for txn spending outputs from multiple wallets

mauricio1802 committed:
[models] Change walletManager::sendFromOutputs to work for txns spending outputs from multiple wallets

mauricio1802 committed:
[models] Fixed logic error in signinTxn

mauricio1802 committed:
[models] Fix error creating genericOutputs in walletManager::sendFromOutputs

mauricio1802 committed:
[models] Add TransactionAPI field to walletManger and updateTransactionAPI method for loading it

mauricio1802 committed:
[models] Asign value to walletManager.TransactionAPI

mauricio1802 committed:
[models] Change walletManger methods SendFromAddresses and SendFromOutputs signature, modify QML to call these methods correctly

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
Merge remote-tracking branch 'simelo_alt/olemis_t150_spend_multiwallet' into olemis_t150_spend_multiwallet

olemis committed:
[ci] refs #234 - In Github feature template merge implementation and description sections

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
[util] refs #150 - Lookup signer by ID in GenericMultiWalletSign

olemis committed:
[util] refs #150 - Break import cycle in util package

olemis committed:
[sky] [errors] [util] refs #150 - Fix tests after merge

- Lookup signer by UID during multi-wallet signing
- Fallback to wallet if signer ID is empty
- Fail if wallet can not sign transactions (e.g. distribution wallets)
- Update test cases after changes for #234 in #254

olemis committed:
[sky] refs #150 - Move test functions added for multi-wallet tests out of sky_test.go

olemis committed:
[util] [models] refs #150 - Add LookupSignServiceForWallet to lookup signer but fall back to wallet if ID empty

- Use it in WalletManager.signTxn to fix error reported in https://travis-ci.org/fibercrypto/FiberCryptoWallet/jobs/615798117#L722-L727
- Replace similar implementation in GenericMultiWalletSign

olemis commented on src/coin/skycoin/models/wallet.go:
```diff
@@ -1057,7 +1094,7 @@ func (wlt *LocalWallet) signSkycoinTxn(txn core.Transaction, pwd core.PasswordRe
 			return nil, err
 		}
-		txnFee = uint64(tmpInt64)
+		//txnFee = uint64(tmpInt64)
```
@stdevMauricio1802 why is this change needed ?

mauricio1802 replied:
@olemis because when the incoming txn is a readableTxn the fee is not needed for create the returned txn

olemis committed:
[sky] refs #150 - Fix linter errors

... reported in https://travis-ci.org/fibercrypto/FiberCryptoWallet/jobs/615808298#L642-L672

olemis committed:
[sky] refs #150 - Fix linter errors in test code

... reported in https://travis-ci.org/fibercrypto/FiberCryptoWallet/jobs/615842606#L642-L658

mauricio1802 committed:
[core] Change signature of core.BlockchainSignService::Sign method for receiv multiple PasswordReaders

mauricio1802 committed:
Merge branch 'olemis_t150_spend_multiwallet' of https://github.com/simelo/FiberCryptoWallet into olemis_t150_spend_multiwallet

mauricio1802 committed:
[core][skycoin] Change signature of methods related to sign txn spending outputs from multiple wallets

mauricio1802 committed:
[core] Update util.GenerecMultiWalletSign to work with multiple passwordReaders

mauricio1802 committed:
[models] Update WalletManager::signTxn method for work with multiple passwords

mauricio1802 committed:
[UI] Update walletIsEncrypted method of SubPageSendAdvanced/Simple components

olemis reviewed (commented):
needs control to input password e.g. for decrypting wallet

mauricio1802 committed:
[models] Create QBridge model for connect calls in backend and frontend

mauricio1802 committed:
[models] Modify walletManager::signTxn to use a bridge for make de passwordReader

mauricio1802 committed:
[models] Add mutex function to QBridge and a property to return the value

mauricio1802 committed:
[models] Fix QBridge and add necessary methods for sharing information

mauricio1802 committed:
[models] Add method to walletManager for signing and broadcastin a txn async

mauricio1802 committed:
[UI] Modify UI for don't show password requested in txn details

mauricio1802 committed:
[ui] Move unlock mutex of bridgeForPassword when the dialog is closed

mauricio1802 committed:
[model] Fix error in signTxn when an incorrect password was provided

mauricio1802 committed:
[models] Remove debug code

mauricio1802 committed:
Merge branch 'fiberDevelop' into stdevMauricio1802_t150_spend_multiwallets

mauricio1802 committed:
Merge branch 'olemis_t150_spend_multiwallet' of https://github.com/simelo/FiberCryptoWallet into stdevMauricio1802_t150_spend_multiwallets

stdevMac committed:
Merge remote-tracking branch 'origin/develop' into olemis_t150_spend_multiwallet

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
[sky] [core] [models] [ui] refs #150 - Rename KeyValueStorage => KeyValueStore

olemis committed:
[mocks] [sky] [core] [util] refs #150 - PasswordReader type receives KeyValueStore with context info

- Coin mocks updated
- New mocks for BlockchainSignService, BlockchainTransactionAPI, KeyValueStore, WalletAddress, WalletOutput
- String constants for wallet signing ops
- Type names for logging
- Rename signingKeyPair.signer => signingKeyPair.signerID
- Rename KeyValueStorage => KeyValueStore
- Add NewMapWithSingleKey
- Add KeyValuesWithDefaults

olemis committed:
[mocks] [sky] [core] [main] [models] [util] refs #150 - Lowercase names in import statements FiberCryptoWallet => fibercryptowallet

olemis committed:
[ci] [models] refs #150 - Update password reader signatures in wallet manager QT models code

olemis committed:
[models] [sky] refs #150 - Retrieve wallet label from pwd reader context for use in QBridge

olemis committed:
[models] [util] refs #150 - ConstantPassword helper function. Use it in wallet manager models.

- Clear plain passwords from memory to reduce the risk of undesired retrieval

olemis committed:
[ci] refs #150 - Lower case project name in +go_import_path

olemis committed:
[ci] refs #150 - Do not convert project name to lowercase during make gen-mocks

olemis committed:
[ci] refs #150 - Travis Linux build does not run in Docker container

olemis committed:
[ci] refs #150 - Install deps before building project on Travis Linux jobs

olemis committed:
[models] refs #150 - Generic type of password reader argument in signTxn

... aimed at fixing error reported in https://travis-ci.org/fibercrypto/fibercryptowallet/jobs/618818243#L768

olemis committed:
[ci] refs #150 - Restore Travis Linux builds inside #Docker container

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis committed:
[sky] [ci] refs #150 - Pass util.EmptyPassword in to makeLocalWalletsFromKeyData

olemis committed:
[ci] refs #150 - Fix syntax error in Travis script after merge

mauricio1802 committed:
[models] Add BeginUse and EndUse methods to QBridge

mauricio1802 committed:
[models]Remove duplicated methods in walletManager

olemis committed:
Merge remote-tracking branch 'fc/develop' into olemis_t150_spend_multiwallet

olemis commented:
@stdevMauricio1802 do I have green light to review and merge ?
